### PR TITLE
FiberIO Directory Support and XML Binary support

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

This PR reworks some of the DASCore discovery internals so that `FiberIO` can now work on directories and implements a directory spool format called `XML_Binary`. `XML_Binary` stores a single xml file at the top of a directory with metadata and the rest of the files are simple binary files. 

TODO:

- [ ] Run profile tests to determine how much this slows down normal indexing and patch retrieval
- [x] @ahmadtourei to test on his datasets which use this format
- [ ] Update contributing docs to mention directory `FiberIO`. 

## Checklist

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [ ] documented the new feature with docstrings or appropriate doc page.
- [ ] included a test. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [ ] your name has been added to the contributors page (docs/contributors.md).
- [ ] added the "ready_for_review" tag once the PR is ready to be reviewed.

### CODE REVIEW

**Comment by d-chambers** [comment]
Hey @ahmadtourei, feel free to take this branch for a spin and let me know what you find.

**Comment by ahmadtourei** [comment]
> Hey @ahmadtourei, feel free to take this branch for a spin and let me know what you find.

Hey @d-chambers, Thanks for prioritizing working on this!
It seems issue #381 still exists and the spool has no patches. The progress bar did not show up (but it does in the master branch). Also, the index_path is in the data directory, but does not exist there.
`dc.spool(data_dir).index.index_path.exists()` raises this error:

```{python}
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[9], [line 1](vscode-notebook-cell:?execution_count=9&line=1)
----> [1](vscode-notebook-cell:?execution_count=9&line=1) dc.spool(data_dir).index.index_path.exists()

AttributeError: 'DirectorySpool' object has no attribute 'index'

```

One other thing to note is when I re-run `sp = dc.spool(data_dir).update()`, it raises the following error (in the master branch it re-index it):
```{python}
---------------------------------------------------------------------------
IsADirectoryError                         Traceback (most recent call last)
File ~/codes/dascore/dascore/utils/hdf5.py:317, in HDFPatchIndexManager._read_metadata(self)
    [316](/codes/dascore/dascore/utils/hdf5.py:316) try:
--> [317](/codes/dascore/dascore/utils/hdf5.py:317)     with _HDF5Store(self.path, "r") as store:
    [318](/codes/dascore/dascore/utils/hdf5.py:318)         out = store.get(self._meta_node)

File ~/codes/dascore/dascore/utils/hdf5.py:78, in _HDF5Store.__init__(self, path, mode, complevel, complib, fletcher32, **kwargs)
     [77](/codes/dascore/dascore/utils/hdf5.py:77) else:
---> [78](/codes/dascore/dascore/utils/hdf5.py:78)     self.open(mode)

File ~/miniconda3/envs/dascore/lib/python3.12/site-packages/pandas/io/pytables.py:745, in HDFStore.open(self, mode, **kwargs)
    [743](/miniconda3/envs/dascore/lib/python3.12/site-packages/pandas/io/pytables.py:743)     raise ValueError(msg)
--> [745](/miniconda3/envs/dascore/lib/python3.12/site-packages/pandas/io/pytables.py:745) self._handle = tables.open_file(self._path, self._mode, **kwargs)

File ~/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:294, in open_file(filename, mode, title, root_uep, filters, **kwargs)
    [293](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:293) # Finally, create the File instance, and return it
--> [294](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:294) return File(filename, mode, title, root_uep, filters, **kwargs)

File ~/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:744, in File.__init__(self, filename, mode, title, root_uep, filters, **kwargs)
    [743](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:743) # Now, it is time to initialize the File extension
--> [744](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:744) self._g_new(filename, mode, **params)
    [746](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/file.py:746) # Check filters and set PyTables format version for new files.

File ~/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/hdf5extension.pyx:394, in tables.hdf5extension.File._g_new()
...
--> [148](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/utils.py:148)     raise IsADirectoryError(f"``{path}`` is not a regular file")
    [149](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/utils.py:149) if not os.access(path, os.R_OK):
    [150](/miniconda3/envs/dascore/lib/python3.12/site-packages/tables/utils.py:150)     raise PermissionError(f"file ``{path}`` exists but it can not be read")

IsADirectoryError: ``/home/data_1`` is not a regular file

```
I'll be at the office this afternoon for further tests. Thanks!

**Comment by d-chambers** [comment]
> I'll be at the office this afternoon for further tests. Thanks!

Sounds good.

> `dc.spool(data_dir).index.index_path.exists()` raises

This raises on the master branch as well. Sorry, I know the indexer stuff is a bit complex. The attribute to access the indexer from the spool is actually `indexer`. So it should be `dc.spool(data_dir).indexer.index_path`. I think I know what's going on with the other issue, but I will need to dig more into it.

**Comment by d-chambers** [comment]
>  Run profile tests to determine how much this slows down normal indexing and patch retrieval

Surprisingly, after "warming up" an external drive with ~12TB of DAS data (~5k files) the indexing on master consistently takes ~31 seconds, while the indexing with this branch consistently takes <~13 seconds.

**Comment by ahmadtourei** [comment]
@d-chambers I just checked and it could index a directory with a bunch of subdirectories containing ~168,000 binary DAS files in less than 500 sec. Also, the data dimensions and coordinates are correct.

**Comment by d-chambers** [comment]
Cool, mind giving the PR a review? Once your happy with it I think it's good to merge.

**Comment by ahmadtourei** [comment]
> Cool, mind giving the PR a review? Once your happy with it I think it's good to merge.

Sounds good! I'll review it in a day or two. Also, please review my last commit and let me know if anything needs to be changed.

File: dascore/viz/wiggle.py [hunk 1]
```diff
@@ -39,8 +39,11 @@ def _shade(offsets, ax, data_scaled, color, wiggle_labels):
 
 def _format_y_axis_ticks(ax, offsets, other_axis_ticks, max_ticks=10):
     """Format the Y axis tick labels."""
+    # make sure not printing too many digits on the figure
+    if other_axis_ticks.dtype != np.dtype("<M8[ns]"):
+        other_axis_ticks = np.around(other_axis_ticks, decimals=2)
```
Comment (d-chambers) [inline dascore/viz/wiggle.py:44]: Probably better to us `dtype_time_like` from utils.time so it will catch other precisions as well.

**Comment by d-chambers** [comment]
Hey @ahmadtourei, 

If you don't have any other issues with the PR I think its ready to merge. It will be much easier to merge this before #375.

**Comment by ahmadtourei** [comment]
> Hey @ahmadtourei,
> 
> If you don't have any other issues with the PR I think its ready to merge. It will be much easier to merge this before #375.

I just need to test one more thing tomorrow. Thanks!

File: tests/test_utils/test_misc.py [hunk 2]
```diff
@@ -300,3 +330,31 @@ def test_nothing(self):
...
+#         path = Path(tmp_path_factory.mktemp('basic_dir'))
+#         (path / 'file.txt').write_text("content of txt")
+#         folder_path = (path / "folder")
+#         folder_path.mkdir(exist_ok=True)
+#         (folder_path / "file.txt").write_text("content of sub text")
+#         (folder_path / "another.bin").write_bytes(b"bytey bytes")
+#         sub_folder_path = folder_path / "sub_folder"
+#         sub_folder_path.mkdir(exist_ok=True)
+#         (sub_folder_path / "file.txt").write_text("content of sub sub text")
+#         return path
```
Comment (ahmadtourei) [inline tests/test_utils/test_misc.py:360]: Do we need this test?

Comment (d-chambers) [reply tests/test_utils/test_misc.py]: No, in fact we don't, I am not sure how I missed this. Good find.

File: dascore/io/xml_binary/__init__.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+"""
+Module for reading data stored in xml_binary format.
+
+This format is a directory which contains metadata as a single xml file
+as well as any number of binary files (raw numeric buffers) which contain
+information about their start time in the file name. An example directory
+might look like this:
+
+data_folder
+   metadata.xml
```
Comment (ahmadtourei) [inline dascore/io/xml_binary/__init__.py:10]: Currently, we only consider XML files named "metadata" so we should note this in the documentation. In the future, we may relax this restriction and look for any XML file in each directory, raising an error if there is more than one.

File: dascore/io/xml_binary/utils.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+def _paths_to_df(paths, metadata, attr_cls):
+    """Convert paths to dataframe of info."""
+    attrs = _paths_to_attrs(paths=paths, metadata=metadata, attr_cls=attr_cls)
+    return _model_list_to_df(attrs)
+
+
+def _read_single_file(path, metadata, time, distance):
+    """Read a single file into a patch."""
+    assert not metadata.transposed_data, "Cant handle data transposition yet."
```
Comment (ahmadtourei) [inline dascore/io/xml_binary/utils.py:155]: Do we want to raise maybe a warning when the number of channels and/or time samples from the XML file is different from the raw binary data.shape? Currently, we read the data based on the XML file, which makes sense, but I think it doesn't hurt to raise a warning.

**Review by ahmadtourei — COMMENTED** [review]
Looks great! Please feel free to merge. Thanks for your help in implementing this IO!

File: dascore/io/xml_binary/utils.py (hunk 4 above)
Comment (d-chambers) [reply dascore/io/xml_binary/utils.py:155]: I just added an assert on the shape being what we expect, I think that's probably the best we can do in this case.

File: dascore/io/xml_binary/__init__.py (hunk 3 above)
Comment (d-chambers) [reply dascore/io/xml_binary/__init__.py:10]: I think we can devise a more general directory-based binary structure once we have support for an inventory-like thing. Let's defer to that point to address this.

**Pull request merged**